packages/leancode_flutter_svg_adaptive_loader/lib/src/leancode_flutter_svg_adaptive_loader.dart

```diff
@@ -0,0 +1,149 @@
+import 'dart:convert';
+import 'dart:math';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_svg/flutter_svg.dart';
+import 'package:vector_graphics/vector_graphics.dart';
+import 'package:vector_graphics_compiler/vector_graphics_compiler.dart' as vg;
+
+/// A [BytesLoader] able to distinguish xml-based svg from compiled binary form,
+/// by looking for '<svg' pattern at 1000 of first characters.
+/// Using logic from [AssetBytesLoader] for binaries and [SvgAssetLoader]
+/// for standard xml-based format.
+class LeancodeFlutterSvgAdaptiveLoader extends BytesLoader {
+  /// Creates a [BytesLoader]
+  const LeancodeFlutterSvgAdaptiveLoader(
+    this.assetName, {
+    this.packageName,
+    this.assetBundle,
+  });
+
+  /// The name of the asset to load.
+  final String assetName;
+
+  /// The package name to load from, if any.
+  final String? packageName;
+
+  /// The asset bundle to use.
+  ///
+  /// If unspecified, [DefaultAssetBundle.of] the current context will be used.
+  final AssetBundle? assetBundle;
+
+  AssetBundle _resolveBundle(BuildContext? context) {
+    if (assetBundle != null) {
+      return assetBundle!;
+    }
+    if (context != null) {
+      return DefaultAssetBundle.of(context);
+    }
+    return rootBundle;
+  }
+
+  Future<ByteData> _prepareMessage(BuildContext? context) {
+    return _resolveBundle(context).load(
+      packageName == null ? assetName : 'packages/$packageName/$assetName',
+    );
+  }
+
+  String _decode(ByteData? message, [int? lenght]) =>
+      utf8.decode(message!.buffer.asUint8List(0, lenght), allowMalformed: true);
+
+  /// This method intentionally avoids using `await` to avoid unnecessary event
+  /// loop turns. This is meant to to help tests in particular.
+  @override
+  Future<ByteData> loadBytes(BuildContext? context) {
+    return _prepareMessage(context).then((message) {
+      // Dart encodes strings in UTF-16 so one codepoint is 2 bytes
+      final first1000Chars = _decode(message, min(2000, message.lengthInBytes));
+
+      if (first1000Chars.contains('<svg')) {
+        return svg.cache.putIfAbsent(
```

Comment:
Where is this `svg` object coming from? `putIfAbsent` sounds like a `Map` API, does it mean we are never controlling cache size and just pushing new caches?

Reply:
That's how the author implemented it in the `flutter_svg` package. 

That's how it looks in flutter_svg implementation

`/// Instance for [Svg]'s utility methods, which can produce a [DrawableRoot]
/// or [PictureInfo] from [String] or [Uint8List].
final Svg svg = Svg._();

/// A utility class for decoding SVG data to a [DrawableRoot] or a [PictureInfo].
///
/// These methods are used by [SvgPicture], but can also be directly used e.g.
/// to create a [DrawableRoot] you manipulate or render to your own [Canvas].
/// Access to this class is provided by the exported [svg] member.
class Svg {
  Svg._();

  /// A global override flag for [SvgPicture.cacheColorFilter].
  ///
  /// If this is null, the value in [SvgPicture.cacheColorFilter] is used. If it
  /// is not null, it will override that value.
  @Deprecated('This no longer does anything.')
  bool? cacheColorFilterOverride;

  /// The cache instance for decoded SVGs.
  final Cache cache = Cache();
}`

Reply:
```

/// The cache for decoded SVGs.
class Cache {
  final Map<Object, Future<ByteData>> _pending = <Object, Future<ByteData>>{};
  final Map<Object, ByteData> _cache = <Object, ByteData>{};

  /// Maximum number of entries to store in the cache.
  ///
  /// Once this many entries have been cached, the least-recently-used entry is
  /// evicted when adding a new entry.
  int get maximumSize => _maximumSize;
  int _maximumSize = 100;

  /// Changes the maximum cache size.
  ///
  /// If the new size is smaller than the current number of elements, the
  /// extraneous elements are evicted immediately. Setting this to zero and then
  /// returning it to its original value will therefore immediately clear the
  /// cache.
  set maximumSize(int value) {
    assert(value != null); // ignore: unnecessary_null_comparison
    assert(value >= 0);
    if (value == maximumSize) {
      return;
    }
    _maximumSize = value;
    if (maximumSize == 0) {
      clear();
    } else {
      while (_cache.length > maximumSize) {
        _cache.remove(_cache.keys.first);
      }
    }
  }

  /// Evicts all entries from the cache.
  ///
  /// This is useful if, for instance, the root asset bundle has been updated
  /// and therefore new images must be obtained.
  void clear() {
    _cache.clear();
  }

  /// Evicts a single entry from the cache, returning true if successful.
  bool evict(Object key) {
    return _cache.remove(key) != null;
  }

  /// Evicts a single entry from the cache if the `oldData` and `newData` are
  /// incompatible.
  ///
  /// For example, if the theme has changed the current color and the picture
  /// uses current color, [evict] will be called.
  bool maybeEvict(Object key, SvgTheme oldData, SvgTheme newData) {
    return evict(key);
  }

  /// Returns the previously cached [PictureStream] for the given key, if available;
  /// if not, calls the given callback to obtain it first. In either case, the
  /// key is moved to the "most recently used" position.
  ///
  /// The arguments must not be null. The `loader` cannot return null.
  Future<ByteData> putIfAbsent(
    Object key,
    Future<ByteData> loader(),
  ) {
    assert(key != null); // ignore: unnecessary_null_comparison
    assert(loader != null); // ignore: unnecessary_null_comparison
    Future<ByteData>? pendingResult = _pending[key];
    if (pendingResult != null) {
      return pendingResult;
    }

    ByteData? result = _cache[key];
    if (result != null) {
      // Remove the provider from the list so that we can put it back in below
      // and thus move it to the end of the list.
      _cache.remove(key);
    } else {
      pendingResult = loader();
      _pending[key] = pendingResult;
      pendingResult.then((ByteData data) {
        _pending.remove(key);
        _add(key, data);
        result = data; // in case it was a synchronous future.
      });
    }
    if (result != null) {
      _add(key, result!);
      return SynchronousFuture<ByteData>(result!);
    }
    assert(_cache.length <= maximumSize);
    return pendingResult!;
  }

  void _add(Object key, ByteData result) {
    if (maximumSize > 0) {
      if (_cache.containsKey(key)) {
        _cache.remove(key); // update LRU.
      } else if (_cache.length == maximumSize && maximumSize > 0) {
        _cache.remove(_cache.keys.first);
      }
      assert(_cache.length < maximumSize);
      _cache[key] = result;
    }
    assert(_cache.length <= maximumSize);
  }

  /// The number of entries in the cache.
  int get count => _cache.length;
}

```